File: src/lira/vocabulary/role/external_dictionary_adapter.py
"""Translates an external dictionary API response into LIRA vocabulary
candidates, insulating core engine data structures against external
JSON schema breakage. Returns every grammatical category the external
source actually supports for the surface form -- never just the first
meaning, and never a fallback NOUN guess when the source gives no
usable evidence: an entry this adapter can't confidently classify
against a real PartOfSpeech member is dropped, not defaulted."""

import re
from typing import Any, Dict, Iterable, List, Optional, Set, Tuple

from lira.value_objects import Code, Identifier, Text

from ..data.external_word_candidate import ExternalWordCandidate
from ..data.part_of_speech import PartOfSpeech
from ..data.source_reference import SourceReference
from ..data.word_lookup_context import WordLookupContext

_EXTERNAL_POS_NAMES = {
    "noun": PartOfSpeech.NOUN,
    "verb": PartOfSpeech.VERB,
    "adjective": PartOfSpeech.ADJECTIVE,
    "adverb": PartOfSpeech.ADVERB,
    "pronoun": PartOfSpeech.PRONOUN,
    "determiner": PartOfSpeech.DETERMINER,
    "preposition": PartOfSpeech.PREPOSITION,
    "conjunction": PartOfSpeech.CONJUNCTION,
    "interjection": PartOfSpeech.INTERJECTION,
    "numeral": PartOfSpeech.NUMERAL,
    "particle": PartOfSpeech.PARTICLE,
    "auxiliary": PartOfSpeech.AUXILIARY,
    "proper noun": PartOfSpeech.PROPER_NOUN,
    "symbol": PartOfSpeech.SYMBOL,
}


class ExternalDictionaryAdapter:
    @classmethod
    def parse_api_payload(cls, payload: Any, context: WordLookupContext, source_uri: str) -> Tuple[ExternalWordCandidate, ...]:
        if not isinstance(payload, list):
            return ()

        candidates: List[ExternalWordCandidate] = []

        for entry in payload:
            if not isinstance(entry, dict):
                continue

            source_word = str(entry.get("word", context.raw_text)).strip()

            meanings = entry.get("meanings", ())
            if not isinstance(meanings, list):
                continue

            for meaning in meanings:
                candidate = cls._parse_meaning(
                    meaning=meaning,
                    source_word=source_word,
                    context=context,
                    source_uri=source_uri,
                )
                if candidate is not None:
                    candidates.append(candidate)

        return cls._deduplicate(candidates)

    @classmethod
    def _parse_meaning(cls, meaning: Any, source_word: str, context: WordLookupContext, source_uri: str) -> Optional[ExternalWordCandidate]:
        if not isinstance(meaning, dict):
            return None

        raw_part_of_speech = str(meaning.get("partOfSpeech", "")).strip().lower()
        part_of_speech = _EXTERNAL_POS_NAMES.get(raw_part_of_speech)
        if part_of_speech is None:
            # No fallback NOUN here -- an entry this adapter can't map
            # to a real PartOfSpeech member is dropped, not guessed at.
            return None

        definitions = meaning.get("definitions", ())
        if not isinstance(definitions, list):
            definitions = ()

        first_definition = next(
            (definition for definition in definitions if isinstance(definition, dict) and definition.get("definition")),
            None,
        )
        definition_text = Text(value=str(first_definition["definition"])) if first_definition else None

        usage_notes = tuple(
            Text(value=str(definition["example"]))
            for definition in definitions
            if isinstance(definition, dict) and definition.get("example")
        )

        domain_relevance = cls._calculate_domain_relevance(
            definition=definition_text.value if definition_text is not None else "",
            domain_name=context.domain_name,
            surrounding_words=(*context.preceding_words, *context.following_words),
        )

        source_reference = SourceReference(
            source_name=Text(value="Free Dictionary API"),
            external_identifier=Identifier(value=f"{source_word}:{raw_part_of_speech}"),
            reference_uri=Identifier(value=source_uri),
        )

        return ExternalWordCandidate(
            text=context.raw_text,
            lexical_form=context.normalised_text,
            normalised_form=context.normalised_text,
            language_code=Code(value="en"),
            part_of_speech=part_of_speech,
            definition=definition_text,
            usage_notes=usage_notes,
            domain_relevance=domain_relevance,
            source_confidence=0.85,
            source_references=(source_reference,),
        )

    @staticmethod
    def _word_terms(text: str) -> Set[str]:
        """Lowercased word tokens, stripped of surrounding punctuation
        (a definition's final word carries a trailing period more often
        than not) and split on hyphens, so "gas-fired" contributes both
        "gas" and "fired" the same way a compound domain name does."""
        return set(re.findall(r"[^\W_]+", text.casefold().replace("-", " ")))

    @classmethod
    def _calculate_domain_relevance(cls, definition: str, domain_name: str, surrounding_words: Iterable[str]) -> float:
        """Simple deterministic ranking only. The domain hint ranks
        externally returned dictionary senses (e.g. "plant" + "Energy
        Power Generation" should rank the power-station sense above the
        biological one); it never invents a new definition or POS, and
        every sense the source actually supports is still returned by
        parse_api_payload -- ranking, not filtering."""
        definition_terms = cls._word_terms(definition)
        domain_terms = cls._word_terms(domain_name)
        context_terms = {term for word in surrounding_words for term in cls._word_terms(word)}

        domain_overlap = len(definition_terms.intersection(domain_terms))
        context_overlap = len(definition_terms.intersection(context_terms))

        score = 0.25 + min(0.50, domain_overlap * 0.15) + min(0.25, context_overlap * 0.10)

        return min(1.0, score)

    @staticmethod
    def _deduplicate(candidates: Iterable[ExternalWordCandidate]) -> Tuple[ExternalWordCandidate, ...]:
        """One candidate per grammatical category -- LIRA models
        different POS as separate Word records, but not a same-form/
        same-POS meaning conflict (that needs explicit sense handling,
        see DictionaryProcessor.register_conflicting_sense), so within
        one external POS category only the highest-ranked candidate
        survives."""
        best_by_category: Dict[PartOfSpeech, ExternalWordCandidate] = {}

        for candidate in candidates:
            existing = best_by_category.get(candidate.part_of_speech)
            if existing is None or candidate.combined_confidence > existing.combined_confidence:
                best_by_category[candidate.part_of_speech] = candidate

        return tuple(sorted(best_by_category.values(), key=lambda candidate: candidate.combined_confidence, reverse=True))

File: prototype/src/lira/vocabulary/role/external_dictionary_adapter.ts
import type { ExternalWordCandidate } from "../data/external_word_candidate";
import { combinedConfidence } from "../data/external_word_candidate";
import { PartOfSpeech } from "../data/enums/part_of_speech";
import type { SourceReference } from "../data/source_reference";
import type { WordLookupContext } from "../data/word_lookup_context";

/** Translates an external dictionary API response into LIRA vocabulary
 * candidates, insulating core engine data structures against external
 * JSON schema breakage. Returns every grammatical category the
 * external source actually supports for the surface form -- never just
 * the first meaning, and never a fallback NOUN guess when the source
 * gives no usable evidence: an entry this adapter can't confidently
 * classify against a real PartOfSpeech member is dropped, not
 * defaulted.
 *
 * Ported from vocabulary/role/external_dictionary_adapter.py. */

const EXTERNAL_POS_NAMES: Record<string, PartOfSpeech> = {
  noun: PartOfSpeech.NOUN,
  verb: PartOfSpeech.VERB,
  adjective: PartOfSpeech.ADJECTIVE,
  adverb: PartOfSpeech.ADVERB,
  pronoun: PartOfSpeech.PRONOUN,
  determiner: PartOfSpeech.DETERMINER,
  preposition: PartOfSpeech.PREPOSITION,
  conjunction: PartOfSpeech.CONJUNCTION,
  interjection: PartOfSpeech.INTERJECTION,
  numeral: PartOfSpeech.NUMERAL,
  particle: PartOfSpeech.PARTICLE,
  auxiliary: PartOfSpeech.AUXILIARY,
  "proper noun": PartOfSpeech.PROPER_NOUN,
  symbol: PartOfSpeech.SYMBOL,
};

interface ApiDefinition {
  definition?: unknown;
  example?: unknown;
}

interface ApiMeaning {
  partOfSpeech?: unknown;
  definitions?: unknown;
}

interface ApiEntry {
  word?: unknown;
  meanings?: unknown;
}

export class ExternalDictionaryAdapter {
  static parseApiPayload(payload: unknown, context: WordLookupContext, sourceUri: string): readonly ExternalWordCandidate[] {
    if (!Array.isArray(payload)) return [];

    const candidates: ExternalWordCandidate[] = [];

    for (const entry of payload as unknown[]) {
      if (typeof entry !== "object" || entry === null) continue;
      const entryRecord = entry as ApiEntry;
      const sourceWord = String(entryRecord.word ?? context.rawText).trim();

      const meanings = entryRecord.meanings;
      if (!Array.isArray(meanings)) continue;

      for (const meaning of meanings as unknown[]) {
        const candidate = this.parseMeaning(meaning, sourceWord, context, sourceUri);
        if (candidate !== undefined) candidates.push(candidate);
      }
    }

    return this.deduplicate(candidates);
  }

  private static parseMeaning(
    meaning: unknown,
    sourceWord: string,
    context: WordLookupContext,
    sourceUri: string,
  ): ExternalWordCandidate | undefined {
    if (typeof meaning !== "object" || meaning === null) return undefined;
    const meaningRecord = meaning as ApiMeaning;

    const rawPartOfSpeech = String(meaningRecord.partOfSpeech ?? "").trim().toLowerCase();
    const partOfSpeech = EXTERNAL_POS_NAMES[rawPartOfSpeech];
    if (partOfSpeech === undefined) {
      // No fallback NOUN here -- an entry this adapter can't map to a
      // real PartOfSpeech member is dropped, not guessed at.
      return undefined;
    }

    const rawDefinitions = meaningRecord.definitions;
    const definitions: ApiDefinition[] = Array.isArray(rawDefinitions) ? (rawDefinitions as ApiDefinition[]) : [];

    const firstDefinition = definitions.find(
      (definition) => typeof definition === "object" && definition !== null && definition.definition,
    );
    const definitionText = firstDefinition?.definition !== undefined
      ? { value: String(firstDefinition.definition) }
      : undefined;

    const usageNotes = definitions
      .filter((definition) => typeof definition === "object" && definition !== null && definition.example)
      .map((definition) => ({ value: String(definition.example) }));

    const domainRelevance = this.calculateDomainRelevance(
      definitionText?.value ?? "",
      context.domainName,
      [...context.precedingWords, ...context.followingWords],
    );

    const sourceReference: SourceReference = {
      sourceName: { value: "Free Dictionary API" },
      externalIdentifier: { value: `${sourceWord}:${rawPartOfSpeech}` },
      referenceUri: { value: sourceUri },
    };

    return {
      text: context.rawText,
      lexicalForm: context.normalisedText,
      normalisedForm: context.normalisedText,
      languageCode: { value: "en" },
      partOfSpeech,
      definition: definitionText,
      usageNotes,
      domainRelevance,
      sourceConfidence: 0.85,
      sourceReferences: [sourceReference],
    };
  }

  /** Lowercased word tokens, stripped of surrounding punctuation (a
   * definition's final word carries a trailing period more often than
   * not) and split on hyphens, so "gas-fired" contributes both "gas"
   * and "fired" the same way a compound domain name does. */
  private static wordTerms(text: string): Set<string> {
    const matches = text.toLowerCase().replace(/-/g, " ").match(/[^\W_]+/g) ?? [];
    return new Set(matches);
  }

  /** Simple deterministic ranking only. The domain hint ranks
   * externally returned dictionary senses (e.g. "plant" + "Energy
   * Power Generation" should rank the power-station sense above the
   * biological one); it never invents a new definition or POS, and
   * every sense the source actually supports is still returned by
   * parseApiPayload -- ranking, not filtering. */
  private static calculateDomainRelevance(definition: string, domainName: string, surroundingWords: readonly string[]): number {
    const definitionTerms = this.wordTerms(definition);
    const domainTerms = this.wordTerms(domainName);
    const contextTerms = new Set(surroundingWords.flatMap((word) => [...this.wordTerms(word)]));

    const domainOverlap = [...definitionTerms].filter((term) => domainTerms.has(term)).length;
    const contextOverlap = [...definitionTerms].filter((term) => contextTerms.has(term)).length;

    const score = 0.25 + Math.min(0.5, domainOverlap * 0.15) + Math.min(0.25, contextOverlap * 0.1);

    return Math.min(1.0, score);
  }

  /** One candidate per grammatical category -- LIRA models different
   * POS as separate Word records, but not a same-form/same-POS meaning
   * conflict (that needs explicit sense handling, see
   * DictionaryProcessor.registerConflictingSense), so within one
   * external POS category only the highest-ranked candidate
   * survives. */
  private static deduplicate(candidates: readonly ExternalWordCandidate[]): readonly ExternalWordCandidate[] {
    const bestByCategory = new Map<PartOfSpeech, ExternalWordCandidate>();

    for (const candidate of candidates) {
      const existing = bestByCategory.get(candidate.partOfSpeech);
      if (existing === undefined || combinedConfidence(candidate) > combinedConfidence(existing)) {
        bestByCategory.set(candidate.partOfSpeech, candidate);
      }
    }

    return [...bestByCategory.values()].sort((a, b) => combinedConfidence(b) - combinedConfidence(a));
  }
}
